### DISCUSSION TITLE
Access `httpApi` endpoint URL dynamically from lambda function

### DISCUSSION THREAD

**Discussion by ryan-token** [serverless/serverless#11009] [Q&A] [answered]
I have several functions that are triggered by an `httpApi` event. However, I don’t know the URL for what those API endpoints will be until after they’ve been deployed the first time.

Is there a way to access the endpoint URL from within a lambda without hardcoding it? i.e. a way to pass whatever that generated URL **will be** into the environment so I can access it from within a function? Or just a better way to think about this/handle it?

**Scenario**: I do some checks on the endpoint itself to set some parameters within the lambda function. Right now I’m hardcoding the endpoint URL (which I can't do until *after* I deploy the endpoint so that I know the URL), which I’d prefer not to do. I could use something like the SSM param store to store the endpoint URLs instead, but that’s also a manual step.

Any ideas for how to best approach this are appreciated!

**Comment by mnapoli** [comment, marked as answer]
Hi Ryan, yes it's possible via a CloudFormation reference (more details here: https://www.serverless.com/framework/docs/providers/aws/guide/resources#aws-cloudformation-resource-reference)

You have to "guess" the ID of the resources created by Serverless Framework (some are referenced in the doc above, but unfortunately not the HTTP API) and then reference them, either via `!Ref`, `!GetAtt` or even `!Sub`.

In your case I think it's:

```yaml
provider:
  name: aws
  environment:
    API_URL: !Sub 'https://${HttpApi}.execute-api.${AWS::Region}.${AWS::URLSuffix}/${sls:stage}/'
```

(inspired from the example at the bottom of https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-resource-httpapi.html#sam-resource-httpapi--examples)

FYI you can check in the generated CloudFormation template (in `.serverless`) that the `HttpApi` resource is indeed your HTTP API.

**Reply by pgrzesik** [reply]
Just to expand on what @mnapoli mentioned - it should be possible to get the whole endpoint of HTTP API via `GetAtt HttpApi.ApiEndpoint`

**Reply by mnapoli** [reply]
Ohh right! I looked in the SAM documentation instead of CloudFormation.

Makes sense: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-apigatewayv2-api.html#aws-resource-apigatewayv2-api-return-values

```yaml
provider:
  name: aws
  environment:
    API_URL: !GetAtt HttpApi.ApiEndpoint
```

**Reply by ryan-token** [reply]
Amazing! Exactly what I need. Thank you for this!
